packages/stylist-macros/src/inline/output/maybe_static.rs

```diff
@@ -0,0 +1,153 @@
+//! an item appears. This information is leverage to provide improved performance and
+//! static caching of parts of the generated output.
+use std::{
+    iter::FromIterator,
+    ops::{BitAnd, BitAndAssign},
+};
+
+use proc_macro2::TokenStream;
+use quote::quote;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum ExpressionContext {
+    // ```
+    // let width = 500;
+    // style! { width: ${width}; }
+    // //               ^^^^^^ dynamic expression, can't wrap style in Lazy
+    // ```
+    Dynamic,
+    // ```
+    // style! { width: 500px; }
+    // //       ------------- everything is static, do wrap style in Lazy
+    // ```
+    Static,
+    // TODO: we can probably avoid a few allocations if we track which parts
+    // of the ast can be constructed statically (with const methods), which is
+    // even stronger than constructing it in the global context in a Lazy.
+    // Should you decide to implement this, keep in mind to change Self::MAX
+    // and adjust the generation of cow-vec tokens. Also check the usages of
+    // MaybeStatic::statick if they can be upgraded to Const.
+    // Const,
+}
+
+impl ExpressionContext {
+    const MAX: Self = Self::Static;
+}
+
+impl BitAnd for ExpressionContext {
+    type Output = Self;
+
+    fn bitand(self, rhs: Self) -> Self::Output {
+        self.min(rhs)
+    }
+}
+
+impl BitAndAssign for ExpressionContext {
+    fn bitand_assign(&mut self, rhs: Self) {
+        *self = *self & rhs
+    }
+}
+
+// Used e.g. to decide whether a sheet can be statically cached in a Lazy or must be
+// created everytime anew.
+pub struct MaybeStatic<T> {
+    value: T,
+    context: ExpressionContext,
+}
+
+impl<T> MaybeStatic<T> {
+    pub fn in_context(context: ExpressionContext, value: T) -> Self {
+        Self { value, context }
+    }
+    pub fn statick(value: T) -> Self {
+        Self::in_context(ExpressionContext::Static, value)
+    }
+    pub fn dynamic(value: T) -> Self {
+        Self::in_context(ExpressionContext::Dynamic, value)
+    }
+    pub fn into_value(self) -> (T, ExpressionContext) {
+        (self.value, self.context)
+    }
+    pub fn flat_map<V>(self, f: impl FnOnce(T) -> MaybeStatic<V>) -> MaybeStatic<V> {
+        let mapped = f(self.value);
+        MaybeStatic {
+            value: mapped.value,
+            context: self.context & mapped.context,
+        }
+    }
+}
+
+impl MaybeStatic<Vec<TokenStream>> {
+    // Get a TokenStream of an expression of type Cow<'_, [typ]>, containing
+    // as elements the expressions form from the Vec<_> in this MaybeStatic.
+    // Depending on the context in which the expression can be expanded,
+    // uses either Cow::Owned or Cow::Borrowed (currently always Cow::Owned).
+    pub fn into_cow_vec_tokens(self, typ: TokenStream) -> MaybeStatic<TokenStream> {
+        let contents = self.value.into_iter();
+        let quoted = quote! {
+            ::std::borrow::Cow::<'_, [#typ]>::Owned(
```

Comment:
This type casting is redundant, you can simply return a vec here and use `.into()` in other structs.

Reply:
Just wanted to make sure to fix the correct type and not run into ambiguous types here, as the method is used in a few different places. I think it's fine, as in the future, the const-context optimized version below will also return a `Cow` and the compiler shouldn't care. Should I add a comment explaining this?

Reply:
I don't think we need to consider a feature that is not yet available.

We don't know what approach is the `&'static []` going to take.

I think it's not possible without using `Lazy` given how currently AST is structured.

Which I haven't tested the performance of having multiple levels of Lazy.

Reply:
I think it shouldn't be too hard if we expose `const fn` constructors from `stylist_core` (instead of using `into` inside the proc macros). It was going to be my next PR, since I think it can save quite a bit of performance.